lucene/core/src/java/org/apache/lucene/search/comparators/DocComparator.java

```diff
@@ -0,0 +1,185 @@
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.search.FieldComparator;
+import org.apache.lucene.search.FilteringLeafFieldComparator;
+import org.apache.lucene.search.LeafFieldComparator;
+import org.apache.lucene.search.Scorable;
+
+import java.io.IOException;
+
+/**
+ * Comparator that sorts by asc _doc
+ */
+public class DocComparator extends FieldComparator<Integer> {
+    private final int[] docIDs;
+    private int topValue;
+    private boolean topValueSet;
+    private boolean reverse = false; // only used to check if skipping functionality should be enabled
+
+    /** Creates a new comparator based on document ids for {@code numHits} */
+    public DocComparator(int numHits) {
+        docIDs = new int[numHits];
+    }
+
+    @Override
+    public int compare(int slot1, int slot2) {
+        // No overflow risk because docIDs are non-negative
+        return docIDs[slot1] - docIDs[slot2];
+    }
+
+
+    @Override
+    public LeafFieldComparator getLeafComparator(LeafReaderContext context) {
+        // TODO: can we "map" our docIDs to the current
+        // reader? saves having to then subtract on every
+        // compare call
+        return new DocLeafComparator(context);
+    }
+
+    @Override
+    public void setTopValue(Integer value) {
+        topValue = value;
+        topValueSet = true;
+    }
+
+    @Override
+    public Integer value(int slot) {
+        return Integer.valueOf(docIDs[slot]);
+    }
+
+    @Override
+    public void setReverse() {
+        reverse = true;
+    }
+
+
+    /**
+     * DocLeafComparator with skipping functionality.
+     * When sort by _doc asc and "after" document is set,
+     * the comparator provides an iterator that can quickly skip to the desired "after" document.
+     */
+    private class DocLeafComparator implements FilteringLeafFieldComparator {
+        private final int docBase;
+        private int bottom;
+
+        private final boolean enableSkipping;
+        private final int minDoc;
+        private final int maxDoc;
+        private DocIdSetIterator topValueIterator; // iterator that starts from topValue
+
+        private boolean iteratorUpdated = false;
+
+        public DocLeafComparator(LeafReaderContext context) {
+            this.docBase = context.docBase;
+            // skipping functionality is enabled if topValue is set and sort is asc
+            this.enableSkipping = topValueSet && reverse == false ? true: false;
+            if (enableSkipping) {
+                this.minDoc = topValue + 1;
+                this.maxDoc = context.reader().maxDoc();
+                this.topValueIterator = DocIdSetIterator.all(maxDoc);
+            } else {
+                this.minDoc = -1;
+                this.maxDoc = -1;
+                this.topValueIterator = null;
+            }
+        }
+
+        @Override
+        public void setBottom(int slot) {
+            bottom = docIDs[slot];
```

Comment:
We should be able to early terminate here if the total hits threshold has been reached. If it's not reached yet, we can early terminate later in `setHitsThresholdReached`.

Reply:
Any query sorted by doc id can early terminate after N matches. That's an important aspect of the optimization since it can be handled by the hits threshold transparently. If there is no `after` value, the threshold should be an upper bound of the number of document that we will collect in the comparator.

Reply:
@jimczi  Thank you for your comment. I would like to clarify something:
> can early terminate after N matches

What's the way to terminate after N matches here? Is it to update an iterator to an empty iterator?

Isn't this termination already handled in TopFieldCollector with the code around `canEarlyTerminate`?  
Is the plan to remove the code around `canEarlyTerminate` in `TopFieldCollector`? Should we do after we also handle the case on early termination with the same index and query sort?

Reply:
> What's the way to terminate after N matches here? Is it to update an iterator to an empty iterator?

I think so, yes. Updating to an empty iterator is what we do for constant score queries for instance. 

> Isn't this termination already handled in TopFieldCollector with the code around canEarlyTerminate?

The code is only for sorted index and while I think that we should move this code in the sort comparator, I agree that it's out of the scope of this PR. Early termination should be handled in the field comparator so that we don't need to add new logic in the main collector.

Reply:
@jimczi Thanks for the feedback. I have added an early termination after N matches for DocComparator in 21de242.

> The code is only for sorted index 

We also have an early termination in a collector based on [_doc order](https://github.com/apache/lucene-solr/blob/master/lucene/core/src/java/org/apache/lucene/search/TopFieldCollector.java#L177-L180).   I can remove this code later once I can make sure that all queries use `DefaultBulkScorer` that uses collector's iterator.

Reply:
Right, sorry I forgot that we added the early termination logic for _doc order in the collector. Maybe that's ok to leave it as it is then. We can revise after we ensure that all bulk scorer uses the collector's iterator ?

Reply:
@jimczi That sounds good to me.  Thank you for the feedback.  Do you have any further comments for this PR?